channeldb/payment_status.go

```diff
@@ -1,31 +1,37 @@
 package channeldb
 
+import "fmt"
+
 // PaymentStatus represent current status of payment.
 type PaymentStatus byte
 
 const (
-	// StatusUnknown is the status where a payment has never been initiated
-	// and hence is unknown.
-	StatusUnknown PaymentStatus = 0
+	// StatusInitiated is the status where a payment has just been
+	// initiated.
+	StatusInitiated PaymentStatus = 1
 
 	// StatusInFlight is the status where a payment has been initiated, but
 	// a response has not been received.
-	StatusInFlight PaymentStatus = 1
+	StatusInFlight PaymentStatus = 2
 
 	// StatusSucceeded is the status where a payment has been initiated and
 	// the payment was completed successfully.
-	StatusSucceeded PaymentStatus = 2
+	StatusSucceeded PaymentStatus = 3
 
 	// StatusFailed is the status where a payment has been initiated and a
 	// failure result has come back.
-	StatusFailed PaymentStatus = 3
+	StatusFailed PaymentStatus = 4
+
+	// StatusError is the status where a payment has conflict states across
+	// its HTLCs without any inflight HTLCs.
+	StatusError PaymentStatus = 5
```

Comment:
I am still wondering how this can happen. Isn't it possible to prevent getting into this state by handling the conflict at a higher level. Maybe when the payment outcome is written into the database?

It feels like a lot of handling is being done for something that is indicative of a bug in lnd? Or is there an error outcome that is not caused by a bug?

Reply:
tbh I either don't know how it could happen. The status of a given MPP is a bit complex as it's determined by multiple non-independent states. My take is we need to precisely mark each state instead of using a vague state to represent it. This way we'd easily know something is wrong.

Reply:
But there must be a db transaction that moves the payment into the error state right? That should be fairly easy to detect using the same logic. Why not return an error right there and refuse to record the invalid state in the first place?

I think it is worth figuring out how this can happen before adding new code to handle something that maybe doesn't need any handling at all if caught early.

Reply:
> But there must be a db transaction that moves the payment into the error state right?

Exactly yeah. I think `decidePaymentStatus` can then return an error and be used as a validation inside all the payment transactions. Good call!

Reply:
Okay that worked out great. No more error state!